apiserver/oauth-proxy-proxy.md

```diff
@@ -0,0 +1,141 @@
+re-used by numerous external teams by directly embedding it into a pod spec.  This changes the tradeoff, so it becomes worthwhile
+to write this code.
+
+### User Stories [optional]
+
+#### Story 1
+
+#### Story 2
+
+### Implementation Details/Notes/Constraints [optional]
+
+1. If HTTP env vars are set, the dynamic proxy-support code must be disabled.  The user is in charge of what the `oauth-proxy`
+ does in these cases.
+2. The list/watch driving the informers must be focused to a single item, so the RBAC special case for single item watch
+ takes effect.
+3. To be performant, the proxy and TLS configuration should be cached in something like a `sync.Atomic`.
+
+
+### Risks and Mitigations
+
+This code is intricate to write, but luckily our monitoring stack already relies on this.  If their tests pass when they
+update to use the feature, the feature works.
+
+## Design Details
+
+### Test Plan
+
+We would get @sur to update the monitoring stack to avoid setting the HTTP env vars and ensure that the proxy e2e test
+still works.
+
+### Graduation Criteria
+
+### Upgrade / Downgrade Strategy
+
+None, fully compatible.
+
+### Version Skew Strategy
+
+None, fully compatible.
+
+## Implementation History
+
+Major milestones in the life cycle of a proposal should be tracked in `Implementation
+History`.
+
+## Drawbacks
+
+The idea is to find the best form of an argument why this enhancement should _not_ be implemented.
+
+## Alternatives
+
+1. Using the internal service DNS name. The .well-known endpoint is public and conformant to a standard RFC.
+ It can only include a single URL, which means it must
+ be public so that external clients like `oc`, the webconsole, and anything else can successfully use it.  Using an internal
+ service name would also preclude the kube-apiserver from ever using an external oauth provider like keycloak.
+
+2. Setting HTTP env vars.  We are doing this is to deliver a mostly working 4.2.0, but has undesireable, long-term characteristics.
+ Setting env vars on the `oauth-proxy` container is possible, but it requires informing every team using the `oauth-proxy`
+ that they need to change.  If they are set, we should honor them, but communicating the scope of this change is difficult; 
+ we can do it for the few in our payload before we ship 4.2.0, but the fanout afterwards is hard.
+
+3. User responsible for MITM proxy.  We don't want to attempt this in 4.2.0.
+ MITM proxies require setting ca-bundles. These come from configmaps which the pod author of `oauth-proxy` also has to know to create.
+ Based on the apiserver team's experience, wiring these ca-bundle requires changes to the `oauth-proxy` to react to changes
+ to the configmaps (so you pick up changes) and bash glue to move the ca-bundle content to the correct location.
+ Communicating these restrictions and capabilities is almost more work than making the `oauth-proxy` reactive with a custom transport.
```

Comment:
isn't there another alternative (implementation alternative anyway) in which the oauth-proxy 

1) launches itself in an init mode which reads the proxy config (and CA) on startup  (in its own gocode/main)
2) writes the CA to /etc/pki/......   
3) execs itself w/ some new arg(not init mode) (because it needs to reinit to load the updated CAs in the TLS libs)
4) watches the proxy config/CA for changes and on changes, kills itself so it goes to (1)?

It's not necessarily simple, but if there are concerns about custom transports, this would seem to address that?  The biggest potential downside is it means there are periods when the oauth-proxy would be "down" during proxy config changes/CA rotations.

At a minimum i think it's worth mentioning why we don't want to pursue this alternative.

Reply:
I don't think we'll pursue something like that. It sounds more fragile to do seamlessly inside of containers you do not own, with valid healthchecks you didn't write.  Remember that this binary is used as part of a user-owned container spec.

Reply:
those reasons make sense to me..  any objection to including the alternative and reason for rejection in the design for historical purposes?